Core/Systems/CutsceneSystem/CutscenePlayer.cs

```diff
@@ -0,0 +1,201 @@
+using System;
+using Terraria.ID;
+
+namespace StarlightRiver.Core.Systems.CutsceneSystem
+{
+	internal class CutscenePlayer : ModPlayer
+	{
+		public Cutscene activeCutscene;
+
+		public int fadeTimer;
+
+		public bool InCutscene => activeCutscene != null;
+
+		/// <summary>
+		/// Sets a given cutscene type as active for this player. This activates cutscene mode.
+		/// </summary>
+		/// <typeparam name="T"></typeparam>
+		public void SetActiveCutscene<T>() where T : Cutscene
+		{
+			activeCutscene?.EndCutscene(Player); // end current cutscene to begin another
+			activeCutscene = (Cutscene)Activator.CreateInstance(typeof(T).Assembly.FullName, typeof(T).FullName).Unwrap();
+
+			if (Main.netMode == NetmodeID.MultiplayerClient)
+			{
+				var packet = new CutscenePacket((byte)Player.whoAmI, typeof(T).FullName, true);
+				packet.Send(-1, Player.whoAmI, false);
+			}
+		}
+
+		/// <summary>
+		/// Sets a given cutscene type as active for this player. This activates cutscene mode.
+		/// 
+		/// This version should typically only be called from net sync packets
+		/// </summary>
+		/// <param name="fullName">The fully qualified name of the cutscene type</param>
+		public void SetActiveCutscene(string fullName)
+		{
+			Mod.Logger.Info($"Activating cutscene {fullName} for {Player.name}");
+			activeCutscene?.EndCutscene(Player); // end current cutscene to begin another
+			activeCutscene = (Cutscene)Activator.CreateInstance(GetType().Assembly.FullName, fullName).Unwrap();
+		}
+
+		/// <summary>
+		/// Deactivates a cutscene if one is active
+		/// </summary>
+		public void DeactivateCutscene()
+		{
+			Mod.Logger.Info($"Ending cutscene for {Player.name}");
+			activeCutscene?.EndCutscene(Player);
+			activeCutscene = null;
+
+			if (Main.netMode == NetmodeID.MultiplayerClient)
+			{
+				var packet = new CutscenePacket((byte)Player.whoAmI, "", false);
+				packet.Send(-1, Player.whoAmI, false);
+			}
+		}
+
+		public override void PreUpdate()
+		{
+			if (InCutscene)
+			{
+				activeCutscene?.UpdateCutscene(Player);
+
+				if (fadeTimer < 60)
+					fadeTimer++;
+			}
+			else
+			{
+				if (fadeTimer > 0)
+					fadeTimer--;
+			}
+		}
+
+		public override void SetControls()
+		{
+			if (InCutscene)
+			{
+				// Disable inputs while in a cutscene
+				Player.controlLeft = false;
+				Player.controlRight = false;
+				Player.controlUp = false;
+				Player.controlDown = false;
+				Player.controlJump = false;
+				Player.controlHook = false;
+				Player.controlInv = false;
+				Player.controlUseItem = false;
+				Player.controlUseTile = false;
+			}
+		}
+	}
+
+	internal class CutsceneSystem : ModSystem
+	{
+		/// <summary>
+		/// This handles the UI 'fadeout' (in reality, the game draws again over the UI with opacity)
+		/// </summary>
+		/// <param name="spriteBatch"></param>
+		public override void PostDrawInterface(SpriteBatch spriteBatch)
+		{
+			CutscenePlayer mp = Main.LocalPlayer.GetModPlayer<CutscenePlayer>();
```

Comment:
I prefer to name the actual modplayer as a variable over mp

Reply:
This has become a pattern across the mod, perhaps a small-scale refactor for it is in order at some point?